src/thor/isochrone.cc

```diff
@@ -200,76 +245,59 @@ void Isochrone::UpdateIsoTile(const EdgeLabel& pred,
   }
 
   // Get the time and distance at the end node of the predecessor
-  // TODO - do we need partial shape from origin location to end of edge?
   float secs1 = pred.cost().secs;
   float dist1 = static_cast<float>(pred.path_distance());
 
   // For short edges just mark the segment between the 2 nodes of the edge. This
   // avoid getting the shape for short edges.
-  if (edge->length() < shape_interval_ * 1.5f) {
-    // Mark tiles that intersect the segment. Optimize this to avoid calling the Intersect
-    // method unless more than 2 tiles are crossed by the segment.
+  auto len = pred.origin() ? pred.path_distance() : edge->length();
+  if (len < shape_interval_ * 1.5f) {
     PointLL ll0 = tile->get_node_ll(t2->directededge(opp)->endnode());
-    auto tile1 = isotile_->TileId(ll0);
-    auto tile2 = isotile_->TileId(ll);
-    if (tile1 == tile2) {
-      isotile_->SetIfLessThan(tile1, {secs1 * kMinPerSec, dist1 * kKmPerMeter});
-    } else if (isotile_->AreNeighbors(tile1, tile2)) {
-      // If tile 2 is directly east, west, north, or south of tile 1 then the
-      // segment will not intersect any other tiles other than tile1 and tile2.
-      isotile_->SetIfLessThan(tile1, {secs1 * kMinPerSec, dist1 * kKmPerMeter});
-      isotile_->SetIfLessThan(tile2, {secs1 * kMinPerSec, dist1 * kKmPerMeter});
-    } else {
-      // Find intersecting tiles (using a Bresenham method)
-      auto tiles = isotile_->Intersect(std::list<PointLL>{ll0, ll});
-      for (const auto& t : tiles) {
-        isotile_->SetIfLessThan(t.first, {secs1 * kMinPerSec, dist1 * kKmPerMeter});
-      }
+    if (pred.origin()) {
+      // interpolate ll0 for origin edge using edge_label.path_distance()
+      auto edge_info = tile->edgeinfo(edge);
+      const auto& shape = edge_info.shape();
+      const auto& ordered_shape =
+          edge->forward() ? shape : std::vector<midgard::PointLL>(shape.rbegin(), shape.rend());
+      auto origin_edge_shape = OriginEdgeShape(ordered_shape, pred.path_distance());
+      ll0 = origin_edge_shape.front();
     }
+    UpdateIsoTileAlongSegment(ll0, ll, secs1, dist1);
     return;
   }
 
   // Get the shape and make sure shape is forward direction. Resample it to
   // the shape interval to get regular spacing. Use the faster resample method.
   // This does not use spherical interpolation - so it is not as accurate but
   // interpolation is over short distances so accuracy should be fine.
-  auto shape = tile->edgeinfo(edge).shape();
-  auto resampled = resample_polyline(shape, edge->length(), shape_interval_);
-  if (!edge->forward()) {
-    std::reverse(resampled.begin(), resampled.end());
+  std::vector<PointLL> resampled;
+  auto edge_info = tile->edgeinfo(edge);
+  const auto& shape = edge_info.shape();
```

Comment:
i see you are taking a const reference here so you cant modify it below. in this case its fine to resample it first and then reverse it and then call just `OriginEdgeShape` right before you use it in the for loop.

but up above where the length is short maybe you can just use the snap point from the origin candidate somehow instead of accessing the shape at all

Reply:
> but up above where the length is short maybe you can just use the snap point from the origin candidate somehow instead of accessing the shape at all

I thought about it. But I haven't found where you can get snap point from `EdgeLabel` struct. Wondering if assume that start and end point is "enough" for short edges and we can just interpolate snap point based on start--end edge rather than fetch whole shape

Reply:
> i see you are taking a const reference here so you cant modify it below. in this case its fine to resample it first and then reverse it and then call just OriginEdgeShape right before you use it in the for loop.

I haven't thought about that! Thanks, applied